main.cpp

```diff
@@ -104,6 +108,406 @@ good to go!
+        value = nullptr;
+    }
+
+
+    IntType& add(int rhs);
+    IntType& subtract(int rhs);
+    IntType& multiply(int rhs);
+    IntType& divide(int rhs);
+
+    IntType& add(const FloatType& rhs);
+    IntType& subtract(const FloatType& rhs);
+    IntType& multiply(const FloatType& rhs);
+    IntType& divide(const FloatType& rhs);
+
+    IntType& add(const DoubleType& rhs);
+    IntType& subtract(const DoubleType& rhs);
+    IntType& multiply(const DoubleType& rhs);
+    IntType& divide(const DoubleType& rhs);
+
+    IntType& add(const IntType& rhs);
+    IntType& subtract(const IntType& rhs);
+    IntType& multiply(const IntType& rhs);
+    IntType& divide(const IntType& rhs);
+};
+
+// ---------
+
+
+FloatType& FloatType::add(float rhs) 
+{
+    if (value != nullptr) *value += rhs;
+    return *this;
+}
+
+FloatType& FloatType::subtract(float rhs) 
+{
+    if (value != nullptr) *value -= rhs;
+    return *this; 
+}
+
+FloatType& FloatType::multiply(float rhs) 
+{
+    if (value != nullptr) *value *= rhs;
+    return *this;
+}
+
+FloatType& FloatType::divide(float rhs)
+{
+    if (rhs == 0.0f) 
+    {
+        std::cout << "warning: floating point division by zero!" << std::endl;
+    }
+    if (value != nullptr) *value /= rhs;
+    return *this;
+}
+
+// ---
+
+FloatType& FloatType::add(const FloatType& rhs)
+{
+   return add(*rhs.value);
+}
+
+FloatType& FloatType::subtract(const FloatType& rhs)
+{
+    return subtract(*rhs.value);
+}
+
+FloatType& FloatType::multiply(const FloatType& rhs)
+{
+    return multiply(*rhs.value);
+}
+
+FloatType& FloatType::divide(const FloatType& rhs)
+{
+    return divide(*rhs.value);
+}
+
+// ---
+
+FloatType& FloatType::add(const DoubleType& rhs)
+{
+   return add(*rhs.value);
+}
+
+FloatType& FloatType::subtract(const DoubleType& rhs)
+{
+    return subtract(*rhs.value);
+}
+
+FloatType& FloatType::multiply(const DoubleType& rhs)
+{
+    return multiply(*rhs.value);
+}
+
+FloatType& FloatType::divide(const DoubleType& rhs)
+{
+    return divide(*rhs.value);
+}
+
+
+// ---
+
+FloatType& FloatType::add(const IntType& rhs)
+{
+   return add(*rhs.value);
+}
+
+FloatType& FloatType::subtract(const IntType& rhs)
+{
+    return FloatType::subtract(*rhs.value);
+}
+
+FloatType& FloatType::multiply(const IntType& rhs)
+{
+    return multiply(*rhs.value);
+}
+
+FloatType& FloatType::divide(const IntType& rhs)
+{
+    return divide(*rhs.value);
+}
+
+// ---
+
+
+DoubleType& DoubleType::add(double rhs)
+{
+    if (value != nullptr) *value += rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::subtract(double rhs)
+{
+    if (value != nullptr) *value -= rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::multiply(double rhs)
+{
+    if (value != nullptr) *value *= rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::divide(double rhs)
+{
+    if (0.0 == rhs) 
+    {
+        std::cout << "warning: floating point division by zero!" << std::endl;
+    }
+    if (value != nullptr) *value /= rhs;
```

Comment:
pick a style with regard to your use of `if( ) single line expression` 
choices:
- use curly braces
- don't use curly braces but put the single-line expression on the next line
- don't use curly braces but put the single-line expression on the same line.
- [ ] revise